File: 프로그래머스/연습문제/Lv2/롤케이크 자르기/롤케이크 자르기.py
from collections import Counter

def solution(topping):
    left = {}
    right = Counter(topping)

    cnt_left,cnt_right = 0,len(right)
    cnt = 0
    for i in range(len(topping)-1):
        right[topping[i]] -= 1
        if right[topping[i]] == 0:
            right.pop(topping[i])
            cnt_right -= 1

        if topping[i] not in left:
            cnt_left += 1
            left[topping[i]] = 1
        else: left[topping[i]] += 1

        if cnt_right == cnt_left: cnt += 1

    return cnt

File: 프로그래머스/연습문제/Lv2/롤케이크 자르기/롤케이크 자르기.cpp
#include <string>
#include <vector>
#include <map>
using namespace std;

int solution(vector<int> topping) {
    map<int, int> left, right;    // <맛 종류, 개수>
    for(int t : topping) right[t]++;
    
    int cnt_left=0, cnt_right=right.size();
    int cnt = 0;
    for(int i=0; i<topping.size()-1; i++){
        // i번째 맛 왼쪽으로 넘기기
        right[topping[i]]--;
        if(right[topping[i]] == 0){
            right.erase(topping[i]);
            cnt_right--;
        }
        
        if(left[topping[i]] == 0) cnt_left++;
        left[topping[i]]++;
        
        if(cnt_right == cnt_left) cnt++;
    }
    
    return cnt;
}
